x/uibc/quota/keeper/quota.go

```diff
@@ -0,0 +1,266 @@
+
+	quotaOfIBCDenom, err := k.getQuotaOfIBCDenom(ctx, denom)
+	if err != nil {
+		return err
+	}
+
+	if quotaOfIBCDenom == nil {
+		return nil
+	}
+
+	amount, ok := sdkmath.NewIntFromString(sendAmount)
+	if !ok {
+		return uibc.ErrInvalidIBCDenom.Wrap("invalid amount")
+	}
+	transferCoin := sdk.NewCoin(denom, amount)
+
+	// convert to base asset if it is `uToken`
+	if ltypes.HasUTokenPrefix(denom) {
+		transferCoin, err = k.leverageKeeper.ExchangeUToken(ctx, transferCoin)
+		if err != nil {
+			return err
+		}
+	}
+
+	// get the exchange price (eg: UMEE) in USD from oracle using base denom eg: `uumee`
+	exchangePrice, err := k.leverageKeeper.TokenValue(ctx, transferCoin, false)
+	if err != nil {
+		// Note: skip the ibc-transfer quota checking if `denom` is not support by leverage
+		if err.Error() == sdkerrors.Wrap(ltypes.ErrNotRegisteredToken, denom).Error() {
+			return nil
+		} else if err != nil {
+			return err
+		}
+	}
+
+	// checking ibc-transfer token quota
+	quotaOfIBCDenom.OutflowSum = quotaOfIBCDenom.OutflowSum.Add(exchangePrice)
+	if quotaOfIBCDenom.OutflowSum.GT(params.TokenQuota) {
+		return uibc.ErrQuotaExceeded
+	}
+
+	// checking total outflow quota
+	totalOutflowSum := k.GetTotalOutflowSum(ctx).Add(exchangePrice)
+	if totalOutflowSum.GT(params.TotalQuota) {
+		return uibc.ErrQuotaExceeded
+	}
+
+	// update the per token outflow sum
+	if err := k.SetQuotaOfIBCDenom(ctx, *quotaOfIBCDenom); err != nil {
+		return err
+	}
+
+	// updating the total outflow sum
+	return k.SetTotalOutflowSum(ctx, totalOutflowSum)
+}
+
+// UndoUpdateQuota undo the quota of ibc denom
+func (k Keeper) UndoUpdateQuota(ctx sdk.Context, denom, sendAmount string) error {
+	quotaOfIBCDenom, err := k.getQuotaOfIBCDenom(ctx, denom)
+	if err != nil {
+		return err
+	}
+
+	if quotaOfIBCDenom == nil {
+		return nil
+	}
+
+	amount, ok := sdkmath.NewIntFromString(sendAmount)
+	if !ok {
+		return uibc.ErrInvalidIBCDenom.Wrap("invalid amount")
+	}
+	transferCoin := sdk.NewCoin(denom, amount)
+
+	// convert to base asset if it is `uToken`
+	if ltypes.HasUTokenPrefix(denom) {
+		transferCoin, err = k.leverageKeeper.ExchangeUToken(ctx, transferCoin)
+		if err != nil {
+			return err
+		}
+	}
+
+	// get the exchange price (eg: UMEE) in USD from oracle using base denom eg: `uumee`
+	exchangePrice, err := k.leverageKeeper.TokenValue(ctx, transferCoin, false)
+	if err != nil {
+		// Note: skip the ibc-transfer quota checking if `denom` is not support by leverage
+		if err.Error() == sdkerrors.Wrap(ltypes.ErrNotRegisteredToken, denom).Error() {
+			return nil
+		} else if err != nil {
+			return err
+		}
+	}
+
+	// reset the outflow limit of token
+	quotaOfIBCDenom.OutflowSum = quotaOfIBCDenom.OutflowSum.Sub(exchangePrice)
+	if err := k.SetQuotaOfIBCDenom(ctx, *quotaOfIBCDenom); err != nil {
+		return err
+	}
+
+	// reset the total outflow sum
+	totalOutflowSum := k.GetTotalOutflowSum(ctx)
+	return k.SetTotalOutflowSum(ctx, totalOutflowSum.Sub(exchangePrice))
+}
```

Comment:
there is some repetition here. Could you possible merge this functions? We could add a param (`undo boolean`) or split it into few functions.